src/systems/ground_reveal.rs

```diff
@@ -0,0 +1,91 @@
+use amethyst::core::specs::prelude::{Read, ReadStorage, System, Write, WriteStorage};
+use amethyst::core::timing::Time;
+use amethyst::core::transform::Transform;
+use amethyst::ecs::Entity;
+
+use entities::tile::LevelGrid;
+use entities::tile::Tile;
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
+use std::time::Duration;
+use systems::TileUpdateQueue;
+
+///
+/// This system handles the recursive process of revealing tiles
+/// including telling the adjacent tiles to update their meshes
+///
+pub struct GroundRevealSystem;
+
+impl<'a> System<'a> for GroundRevealSystem {
+    type SystemData = (
+        Read<'a, Time>,
+        Read<'a, LevelGrid>,
+        Write<'a, BinaryHeap<Reverse<(Duration, Entity)>>>,
+        ReadStorage<'a, Transform>,
+        WriteStorage<'a, Tile>,
+        Write<'a, TileUpdateQueue>,
+    );
+
+    fn run(
+        &mut self,
+        (time, grid, mut heap, transforms, mut tiles, mut tile_update_queue): Self::SystemData,
+    ) {
+        if let Some(Reverse((mut reveal_time, mut entity))) = heap.peek().cloned() {
+            while reveal_time <= time.absolute_time() {
```

Comment:
if i understand correctly, you want to loop as long as the reveal time on the top of the heap is smaller than the current time. if that's correct, i would suggest to do just that:

```
while !heap.is_empty() && heap.peek().unwrap().reveal_time <= time.absolute_time() {
    let Reverse((reveal_time, entity)) = heap.pop().unwrap();
```
not entirely sure whether that works, but that would mean less mut, less clones, and less code.

Reply:
works